dace/transformation/dataflow/hbm_transform.py

```diff
@@ -0,0 +1,198 @@
+                state.add_edge(
+                    other_node, path[0].src_conn, fwtasklet, "_in",
+                    memlet.Memlet(other_node.data, subset=target_other_subset))
+            else:
+                inner_edge = state.add_edge(inner_edge.src, inner_edge.src_conn,
+                                            fwtasklet, '_in', mem)
+                state.add_edge(
+                    fwtasklet, "_out", other_node, path[-1].dst_conn,
+                    memlet.Memlet(other_node.data, subset=target_other_subset))
+
+        utils.update_path_subsets(state, inner_edge, new_subset)
+
+    def _update_array_hbm(self, array_name: str, new_memory: str, new_bank: str,
+                          sdfg: SDFG):
+        """
+        Updates bank assignments for the array
+        """
+        desc = sdfg.arrays[array_name]
+        old_memory = None
+        if 'memorytype' in desc.location and desc.location[
+                "memorytype"] is not None:
+            old_memory = desc.location["memorytype"]
+        if new_memory == "HBM":
+            low, high = fpga.get_multibank_ranges_from_subset(new_bank, sdfg)
+        else:
+            low, high = int(new_bank), int(new_bank) + 1
+        if (old_memory is None or old_memory == "DDR") and new_memory == "HBM":
+            desc = utils.update_array_shape(sdfg, array_name,
+                                            (high - low, *desc.shape))
+        elif old_memory == "HBM" and (new_memory == "DDR"
+                                      or new_memory is None):
+            desc = utils.update_array_shape(array_name, *(list(desc.shape)[1:]))
+        elif old_memory == "HBM" and new_memory == "HBM":
+            new_shape = list(desc.shape)
+            new_shape[0] = high - low
+            desc = utils.update_array_shape(sdfg, array_name, new_shape)
+        desc.location["memorytype"] = new_memory
+        desc.location['bank'] = new_bank
+        desc.storage = dtypes.StorageType.FPGA_Global
+
+    @staticmethod
+    def can_be_applied(self, graph: Union[SDFG, SDFGState],
+                       candidate: Dict['PatternNode', int], expr_index: int,
+                       sdfg: SDFG, strict: bool) -> bool:
+        return True
+
+    @staticmethod
+    def expressions():
+        # Matches anything
+        return [networkx.DiGraph()]
+
+    def apply(self, sdfg: SDFG) -> Union[Any, None]:
+        # update array bank positions
+        for array_name, memory_type, bank in self.update_array_banks:
+            self._update_array_hbm(array_name, memory_type, bank, sdfg)
+
+        # nest the sdfg and execute in parallel
+        if (0, 0, 1) != subsets.Range.from_string(self.outer_map_range[1])[0]:
+            self._multiply_sdfg_executions(sdfg)
+
+        # set default on all outer arrays, such that FPGA_transformation can be used
+        for desc in sdfg.arrays.items():
+            desc[1].storage = dtypes.StorageType.Default
```

Comment:
Can this transformation only be applied _before_ `FPGATransform`? I know that this is the case for some transformations, but if yes then please explain why

Reply:
Yes, or at least it wouldn't make any sense to run it after as far as I can see